internals-js/src/specs/sourceSpec.ts

```diff
@@ -0,0 +1,146 @@
+    const JSONSelection = this.addScalarType(schema, 'JSONSelection');
+    const JSON = this.addScalarType(schema, 'JSON');
+
+    const HTTPSourceType = schema.addType(new InputObjectType('HTTPSourceType'));
+    HTTPSourceType.addField(new InputFieldDefinition('GET')).type = URLPathTemplate;
+    HTTPSourceType.addField(new InputFieldDefinition('POST')).type = URLPathTemplate;
+    HTTPSourceType.addField(new InputFieldDefinition('headers')).type =
+      new ListType(new NonNullType(HTTPHeaderMapping));
+    HTTPSourceType.addField(new InputFieldDefinition('body')).type = JSONSelection;
+    sourceType.addArgument('http', HTTPSourceType);
+
+    sourceType.addArgument('selection', new NonNullType(JSONSelection));
+
+    const KeyTypeMap = schema.addType(new InputObjectType('KeyTypeMap'));
+    KeyTypeMap.addField(new InputFieldDefinition('key')).type = new NonNullType(schema.stringType());
+    KeyTypeMap.addField(new InputFieldDefinition('typeMap')).type = JSON;
+    sourceType.addArgument('keyTypeMap', KeyTypeMap);
+
+    const sourceField = this.addDirective(schema, 'sourceField').addLocations(
+      DirectiveLocation.FIELD_DEFINITION,
+    );
+    sourceField.repeatable = true;
+    sourceField.addArgument('api', new NonNullType(schema.stringType()));
+    sourceField.addArgument('selection', JSONSelection);
+
+    const HTTPSourceField = schema.addType(new InputObjectType('HTTPSourceField'));
+    HTTPSourceField.addField(new InputFieldDefinition('GET')).type = URLPathTemplate;
+    HTTPSourceField.addField(new InputFieldDefinition('POST')).type = URLPathTemplate;
+    HTTPSourceField.addField(new InputFieldDefinition('PUT')).type = URLPathTemplate;
+    HTTPSourceField.addField(new InputFieldDefinition('PATCH')).type = URLPathTemplate;
+    HTTPSourceField.addField(new InputFieldDefinition('DELETE')).type = URLPathTemplate;
+    HTTPSourceField.addField(new InputFieldDefinition('body')).type = JSONSelection;
+    HTTPSourceField.addField(new InputFieldDefinition('headers')).type =
+      new ListType(new NonNullType(HTTPHeaderMapping));
+    sourceField.addArgument('http', HTTPSourceField);
+
+    return [];
+  }
+
+  allElementNames(): string[] {
+    return [
+      '@sourceAPI',
+      '@sourceType',
+      '@sourceField',
+      // 'JSONSelection',
+      // 'URLPathTemplate',
+      // 'JSON',
+      // 'HTTPHeaderMapping',
+      // 'HTTPSourceAPI',
+      // 'HTTPSourceType',
+      // 'HTTPSourceField',
+      // 'KeyTypeMap',
+    ];
+  }
+
+  sourceAPIDirective(schema: Schema) {
+    return this.directive(schema, 'sourceAPI')!;
+  }
+
+  sourceTypeDirective(schema: Schema) {
+    return this.directive(schema, 'sourceType')!;
+  }
+
+  sourceFieldDirective(schema: Schema) {
+    return this.directive(schema, 'sourceField')!;
+  }
+}
+
+export const SOURCE_VERSIONS = new FeatureDefinitions<SourceSpecDefinition>(sourceIdentity)
+  .add(new SourceSpecDefinition(new FeatureVersion(0, 1)));
```

Comment:
we want a second argument on the SourceSpecDefinition that will tie this to a specific federation version. When we release, this will be `2.7`, but we haven't bumped elsewhere yet.

Reply:
https://github.com/apollographql/federation/pull/2894/commits/d10a518a80146346d5213681adafe91ed2f56542